pkg/multitenant/cost_controlling_io.go

```diff
@@ -0,0 +1,208 @@
+		return s
+	}
+	return &accountingExternalStorage{
+		ExternalStorage: s,
+		limit:           limit,
+		recorder:        recorder,
+	}
+}
+
+func (a *accountingExternalStorage) Writer(
+	ctx context.Context, basename string,
+) (io.WriteCloser, error) {
+	inner, err := a.ExternalStorage.Writer(ctx, basename)
+	if err != nil {
+		return nil, err
+	}
+	return &accountingWriter{
+		ctx:      ctx,
+		inner:    inner,
+		limit:    a.limit,
+		recorder: a.recorder,
+	}, nil
+}
+
+func (a *accountingExternalStorage) ReadFile(
+	ctx context.Context, basename string,
+) (ioctx.ReadCloserCtx, error) {
+	inner, err := a.ExternalStorage.ReadFile(ctx, basename)
+	if err != nil {
+		return nil, err
+	}
+	return &accountingReader{
+		inner:    inner,
+		limit:    a.limit,
+		recorder: a.recorder,
+	}, nil
+}
+
+func (a *accountingExternalStorage) ReadFileAt(
+	ctx context.Context, basename string, offset int64,
+) (ioctx.ReadCloserCtx, int64, error) {
+	inner, sz, err := a.ExternalStorage.ReadFileAt(ctx, basename, offset)
+	if err != nil {
+		return nil, 0, err
+	}
+	return &accountingReader{
+		inner:    inner,
+		limit:    a.limit,
+		recorder: a.recorder,
+	}, sz, nil
+}
+
+// accountingWriter is an io.WriteCloser that tracks how many total bytes have been written. If limit is > 0, then the
+// writer will record the written bytes and wait for the associated RUs in a Write call if more than limit bytes have been
+// written. On Close, any previously unaccounted for RUs will be recorded.
+//
+// If limit <= 0 then we will wait for RUs only on Close().
+//
+// NB: The implementation allows roughly 2x the limit to be unaccounted if the caller is making write calls with a large
+// values just under the limit.
+type accountingWriter struct {
+	ctx      context.Context
+	inner    io.WriteCloser
+	recorder TenantSideExternalIORecorder
+	limit    int64
+
+	count int64
+}
+
+var _ io.WriteCloser = (*accountingWriter)(nil)
+
+func (aw *accountingWriter) Write(d []byte) (int, error) {
+	// If past writes have pushed us past the limit, account for them before allowing this write.
+	aw.maybeWaitForRUs()
+
+	// If this single write is larger than the limit, immediately account for it.
+	if int64(len(d)) > aw.limit {
+		return aw.immediatelyAccountedWrite(d)
+	}
+
+	n, err := aw.inner.Write(d)
+	aw.count += int64(n)
+	return n, err
+}
+
+func (aw *accountingWriter) immediatelyAccountedWrite(d []byte) (int, error) {
+	writeLen := int64(len(d))
+	aw.recorder.ExternalIOWriteWait(aw.ctx, writeLen)
+	n, err := aw.inner.Write(d)
+	if err != nil {
+		aw.recorder.ExternalIOWriteFailure(aw.ctx, int64(n), writeLen-int64(n))
+		return n, err
+	}
+	aw.recorder.ExternalIOWriteSuccess(aw.ctx, int64(n))
+	return n, err
+}
+
+// Close closes the underlying Writer and also waits for any RUs that weren't accounted for on a
+// previous call to Write.
+func (aw *accountingWriter) Close() error {
+	aw.recorder.ExternalIOWriteSuccess(aw.ctx, aw.count)
```

Comment:
a bit strange seeing "WriteSuccess" here because close should be invoked even if write failed.

Reply:
I'll add a comment to point out that `count` are the bytes that we've previously successfully written.